File: src/android/FaceDetectorHelper.java
/**
 * FaceDetectorHelper.java
 *
 * Uses ML Kit Face Detection (on-device) to detect and extract portrait images
 * from driver license photos. Falls back to deterministic cropping if no face is detected.
 *
 * Parameters tuned to match iOS Vision framework implementation:
 * - MIN_FACE_SIZE_RATIO: 0.002 (license faces are small in the camera frame)
 * - FACE_PADDING_FACTOR: 0.35 (natural portrait crop around detected face)
 * - Scoring: 70% size / 30% position (larger faces prioritized)
 * - Expected position: 35%/50% (left-center of frame for US licenses)
 */
package com.sos.driverslicensescanner;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Matrix;
import android.graphics.Rect;
import android.util.Log;

import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.mlkit.vision.common.InputImage;
import com.google.mlkit.vision.face.Face;
import com.google.mlkit.vision.face.FaceDetection;
import com.google.mlkit.vision.face.FaceDetector;
import com.google.mlkit.vision.face.FaceDetectorOptions;

import java.util.List;

/**
 * Helper class for detecting faces and extracting portrait images from license photos.
 */
public class FaceDetectorHelper {

    private static final String TAG = "FaceDetectorHelper";

    // Padding factor around detected face (as percentage of face size)
    // Matches iOS kFacePaddingFactor = 0.35
    private static final float FACE_PADDING_FACTOR = 0.35f;

    // Minimum face size relative to image area.
    // A face on a license held at scanning distance is typically 0.3-1% of the full
    // camera frame, so we set a low threshold to avoid rejecting real license faces.
    // Matches iOS kMinFaceSizeRatio = 0.002
    private static final float MIN_FACE_SIZE_RATIO = 0.002f;

    // Maximum face size relative to image (faces larger than this are likely false positives)
    private static final float MAX_FACE_SIZE_RATIO = 0.7f;

    /**
     * Callback interface for face detection results (from captured photo).
     */
    public interface FaceDetectionCallback {
        void onFaceDetected(Bitmap croppedFace);
        void onNoFaceDetected();
        void onFaceDetectionError(String error);
    }

    /**
     * Callback interface for live video face presence detection.
     * Used during front scan to trigger auto-capture when a face is visible.
     * croppedFace is non-null only when faceFound=true and a valid crop could be extracted.
     */
    public interface LiveFaceCallback {
        void onFacePresenceDetected(boolean faceFound, Bitmap croppedFace);
    }

    private final Context context;
    private final FaceDetectionCallback callback;
    private FaceDetector faceDetector;
    private FaceDetector liveFaceDetector;

    /**
     * Creates a new FaceDetectorHelper.
     *
     * @param context The application context
     * @param callback Callback for face detection results
     */
    public FaceDetectorHelper(Context context, FaceDetectionCallback callback) {
        this.context = context;
        this.callback = callback;
        initializeDetector();
        initializeLiveDetector();
    }

    /**
     * Initializes the ML Kit face detector with optimal settings for ID photo portraits.
     */
    private void initializeDetector() {
        FaceDetectorOptions options = new FaceDetectorOptions.Builder()
                .setPerformanceMode(FaceDetectorOptions.PERFORMANCE_MODE_ACCURATE)
                .setLandmarkMode(FaceDetectorOptions.LANDMARK_MODE_NONE)
                .setContourMode(FaceDetectorOptions.CONTOUR_MODE_NONE)
                .setClassificationMode(FaceDetectorOptions.CLASSIFICATION_MODE_NONE)
                // Matches iOS minFaceSize = 0.05 (proportion of image width)
                .setMinFaceSize(0.05f)
                .enableTracking()
                .build();

        faceDetector = FaceDetection.getClient(options);
        Log.d(TAG, "Face detector initialized");
    }

    /**
     * Initializes a face detector for live video frame analysis.
     * Uses PERFORMANCE_MODE_ACCURATE with a very small minFaceSize to
     * detect the small printed face on a driver's license held at arm's length.
     */
    private void initializeLiveDetector() {
        FaceDetectorOptions options = new FaceDetectorOptions.Builder()
                .setPerformanceMode(FaceDetectorOptions.PERFORMANCE_MODE_ACCURATE)
                .setLandmarkMode(FaceDetectorOptions.LANDMARK_MODE_NONE)
                .setContourMode(FaceDetectorOptions.CONTOUR_MODE_NONE)
                .setClassificationMode(FaceDetectorOptions.CLASSIFICATION_MODE_NONE)
                // Very small minFaceSize to detect printed license photos
                .setMinFaceSize(0.01f)
                .build();

        liveFaceDetector = FaceDetection.getClient(options);
        Log.d(TAG, "Live face detector initialized (accurate mode, minFace=0.01)");
    }

    /**
     * Detects a face in the given bitmap and extracts the portrait.
     * Used for captured photos (high-quality, accurate mode).
     *
     * @param bitmap The image to analyze (typically the front of a driver license)
     */
    public void detectFace(Bitmap bitmap) {
        if (bitmap == null || bitmap.isRecycled()) {
            Log.w(TAG, "Invalid bitmap provided");
            callback.onNoFaceDetected();
            return;
        }

        InputImage inputImage = InputImage.fromBitmap(bitmap, 0);

        faceDetector.process(inputImage)
                .addOnSuccessListener(faces -> handleDetectionResult(bitmap, faces))
                .addOnFailureListener(e -> {
                    Log.e(TAG, "Face detection failed", e);
                    callback.onFaceDetectionError(e.getMessage());
                });
    }

    /**
     * Checks for face presence in a live video frame.
     * Uses fast detection mode suitable for real-time video analysis.
     * Does not extract/crop the face — only reports presence.
     *
     * @param bitmap Video frame to analyze
     * @param liveFaceCallback Callback for face presence result
     */
    public void checkForFacePresence(Bitmap bitmap, LiveFaceCallback liveFaceCallback) {
        if (bitmap == null || bitmap.isRecycled()) {
            liveFaceCallback.onFacePresenceDetected(false, null);
            return;
        }

        Log.d(TAG, "checkForFacePresence: bitmap " + bitmap.getWidth() + "x" + bitmap.getHeight());

        InputImage inputImage = InputImage.fromBitmap(bitmap, 0);

        liveFaceDetector.process(inputImage)
                .addOnSuccessListener(faces -> {
                    boolean found = faces != null && !faces.isEmpty();
                    Log.d(TAG, "Live face check result: " + (found ? faces.size() + " face(s) FOUND" : "no faces"));
                    Bitmap cropped = null;
                    if (found) {
                        // Extract the portrait crop right here using the same bitmap and bounding boxes.
                        // This avoids a second detection pass with a different detector later.
                        Face bestFace = findBestFace(bitmap, faces);
                        if (bestFace != null) {
                            cropped = extractFaceRegion(bitmap, bestFace);
                            Log.d(TAG, "Live crop result: " + (cropped != null
                                    ? cropped.getWidth() + "x" + cropped.getHeight() : "null"));
                        }
                    }
                    liveFaceCallback.onFacePresenceDetected(found, cropped);
                })
                .addOnFailureListener(e -> {
                    Log.w(TAG, "Live face detection failed: " + e.getMessage(), e);
                    liveFaceCallback.onFacePresenceDetected(false, null);
                });
    }

    /**
     * Handles the face detection result from captured photo.
     */
    private void handleDetectionResult(Bitmap originalBitmap, List<Face> faces) {
        if (faces == null || faces.isEmpty()) {
            Log.d(TAG, "No faces detected");
            callback.onNoFaceDetected();
            return;
        }

        Log.d(TAG, "Detected " + faces.size() + " face(s)");

        Face bestFace = findBestFace(originalBitmap, faces);

        if (bestFace == null) {
            Log.d(TAG, "No suitable face found within expected parameters");
            callback.onNoFaceDetected();
            return;
        }

        try {
            Bitmap croppedFace = extractFaceRegion(originalBitmap, bestFace);
            if (croppedFace != null) {
                Log.d(TAG, "Successfully extracted face region");
                callback.onFaceDetected(croppedFace);
            } else {
                callback.onNoFaceDetected();
            }
        } catch (Exception e) {
            Log.e(TAG, "Error extracting face region", e);
            callback.onFaceDetectionError(e.getMessage());
        }
    }

    /**
     * Finds the best face candidate from the detected faces.
     * Scoring: 70% size + 30% position (matches iOS implementation).
     * US licenses typically have the photo on the left side.
     */
    private Face findBestFace(Bitmap bitmap, List<Face> faces) {
        int imageWidth = bitmap.getWidth();
        int imageHeight = bitmap.getHeight();
        float imageArea = imageWidth * imageHeight;

        Face bestFace = null;
        float bestScore = 0;

        for (Face face : faces) {
            Rect bounds = face.getBoundingBox();
            float faceArea = bounds.width() * bounds.height();
            float areaRatio = faceArea / imageArea;

            // Skip faces that are too small or too large
            if (areaRatio < MIN_FACE_SIZE_RATIO || areaRatio > MAX_FACE_SIZE_RATIO) {
                Log.d(TAG, "Face rejected due to size: ratio=" + areaRatio);
                continue;
            }

            // Size score (70% weight) — larger faces prioritized
            float sizeScore = faceArea / imageArea;

            // Position score (30% weight) — prefer left-center area
            // Matches iOS: expected position at 35%/50% of image
            float expectedCenterX = imageWidth * 0.35f;
            float expectedCenterY = imageHeight * 0.50f;
            float actualCenterX = bounds.centerX();
            float actualCenterY = bounds.centerY();

            float positionScore = 1.0f - (
                    Math.abs(actualCenterX - expectedCenterX) / imageWidth +
                            Math.abs(actualCenterY - expectedCenterY) / imageHeight
            ) / 2;
            positionScore = Math.max(0, positionScore);

            // Combined score: 70/30 split (matches iOS)
            float totalScore = sizeScore * 0.7f + positionScore * 0.3f;

            Log.d(TAG, String.format("Face score: size=%.4f pos=%.4f total=%.4f (center=%.0f,%.0f)",
                    sizeScore, positionScore, totalScore, actualCenterX, actualCenterY));

            if (totalScore > bestScore) {
                bestScore = totalScore;
                bestFace = face;
            }
        }

        return bestFace;
    }

    /**
     * Extracts the face region from the image with 35% padding.
     */
    private Bitmap extractFaceRegion(Bitmap bitmap, Face face) {
        Rect bounds = face.getBoundingBox();

        float faceCX = bounds.exactCenterX();
        float faceCY = bounds.exactCenterY();
        float faceW  = bounds.width();
        float faceH  = bounds.height();

        float angleZ = face.getHeadEulerAngleZ();
        Log.d(TAG, "Face eulerAngleZ=" + angleZ + "° centre=(" + faceCX + "," + faceCY + ")");

        // Desired output square: face size + portrait padding on all sides
        float faceSize   = Math.max(faceW, faceH);
        float outputSize = (float) Math.ceil(faceSize * (1.0f + FACE_PADDING_FACTOR * 2.0f));

        // Pre-rotation crop must be large enough so rotating it by angleZ leaves
        // no empty corners inside the central outputSize × outputSize region.
        // Required side = outputSize × (|cos| + |sin|).
        double rad = Math.toRadians(angleZ);
        float sinA = (float) Math.abs(Math.sin(rad));
        float cosA = (float) Math.abs(Math.cos(rad));
        float preCropSize = (float) Math.ceil(outputSize * (cosA + sinA)) + 4;

        // Crop centred on the face, clamped to bitmap bounds
        int left   = Math.max(0,                 (int)(faceCX - preCropSize / 2));
        int top    = Math.max(0,                 (int)(faceCY - preCropSize / 2));
        int right  = Math.min(bitmap.getWidth(),  (int)(faceCX + preCropSize / 2));
        int bottom = Math.min(bitmap.getHeight(), (int)(faceCY + preCropSize / 2));

        int w = right - left;
        int h = bottom - top;
        if (w <= 0 || h <= 0) {
            Log.w(TAG, "Invalid pre-crop region");
            return null;
        }

        try {
            Bitmap large = Bitmap.createBitmap(bitmap, left, top, w, h);

            // Rotate the large region to align the face upright.
            // Matrix.setRotate positive = clockwise.
            // angleZ > 0 = face tilted counter-clockwise → clockwise correction = +angleZ.
            Bitmap rotated = (Math.abs(angleZ) >= 3.0f)
                    ? rotateBitmap(large, angleZ)
                    : large;

            // Crop a square from the centre — face is centred, no white corners.
            return centerSquareCrop(rotated, (int) outputSize);

        } catch (Exception e) {
            Log.e(TAG, "Error extracting face region", e);
            return null;
        }
    }

    /**
     * Rotates a bitmap by the given angle (degrees) around its centre.
     * The output canvas expands to contain the full rotated content.
     */
    private Bitmap rotateBitmap(Bitmap src, float angleDegrees) {
        double rad = Math.toRadians(angleDegrees);
        float sinA = (float) Math.abs(Math.sin(rad));
        float cosA = (float) Math.abs(Math.cos(rad));
        int newWidth  = Math.round(src.getWidth() * cosA + src.getHeight() * sinA);
        int newHeight = Math.round(src.getWidth() * sinA + src.getHeight() * cosA);

        Matrix matrix = new Matrix();
        matrix.setRotate(angleDegrees, src.getWidth() / 2f, src.getHeight() / 2f);
        matrix.postTranslate((newWidth - src.getWidth()) / 2f, (newHeight - src.getHeight()) / 2f);

        Bitmap result = Bitmap.createBitmap(newWidth, newHeight, Bitmap.Config.ARGB_8888);
        android.graphics.Canvas canvas = new android.graphics.Canvas(result);
        canvas.drawBitmap(src, matrix, null);
        src.recycle();
        return result;
    }

    /**
     * Crops a square of `size` pixels from the centre of `src`.
     * The centre of `src` is the (corrected) face centre.
     */
    private Bitmap centerSquareCrop(Bitmap src, int size) {
        int side = Math.min(size, Math.min(src.getWidth(), src.getHeight()));
        int x = (src.getWidth()  - side) / 2;
        int y = (src.getHeight() - side) / 2;
        Bitmap result = Bitmap.createBitmap(src, x, y, side, side);
        src.recycle();
        return result;
    }

    /**
     * Performs deterministic cropping based on standard driver license layout.
     * Used as fallback when face detection fails.
     * Percentages match iOS implementation: Y=12%, W=26%, H=55%.
     *
     * @param bitmap The front of the driver license
     * @return Cropped portrait region
     */
    public static Bitmap deterministicCrop(Bitmap bitmap) {
        if (bitmap == null || bitmap.isRecycled()) {
            return null;
        }

        int width = bitmap.getWidth();
        int height = bitmap.getHeight();

        // Matches iOS: 3% left, 12% top, 26% width, 55% height
        int portraitLeft = (int) (width * 0.03);
        int portraitTop = (int) (height * 0.12);
        int portraitWidth = (int) (width * 0.26);
        int portraitHeight = (int) (height * 0.55);

        portraitLeft = Math.max(0, portraitLeft);
        portraitTop = Math.max(0, portraitTop);
        portraitWidth = Math.min(portraitWidth, width - portraitLeft);
        portraitHeight = Math.min(portraitHeight, height - portraitTop);

        if (portraitWidth <= 0 || portraitHeight <= 0) {
            Log.w(TAG, "Invalid deterministic crop region");
            return null;
        }

        try {
            return Bitmap.createBitmap(bitmap, portraitLeft, portraitTop, portraitWidth, portraitHeight);
        } catch (Exception e) {
            Log.e(TAG, "Error performing deterministic crop", e);
            return null;
        }
    }

    /**
     * Releases resources used by the face detectors.
     */
    public void shutdown() {
        if (faceDetector != null) {
            faceDetector.close();
            faceDetector = null;
            Log.d(TAG, "Face detector shut down");
        }
        if (liveFaceDetector != null) {
            liveFaceDetector.close();
            liveFaceDetector = null;
            Log.d(TAG, "Live face detector shut down");
        }
    }
}

File: src/ios/FaceDetectorHelper.m
/**
 * FaceDetectorHelper.m
 *
 * Vision framework face detection implementation for portrait extraction.
 * Detects face on a driver license photo and crops it with padding.
 */

#import "FaceDetectorHelper.h"
#import <Vision/Vision.h>

// Padding factor around detected face for the crop
static CGFloat const kFacePaddingFactor = 0.35;

// Minimum face area relative to image area.
// A face on a license held at scanning distance is typically 0.3-1% of the full
// camera frame, so we set a low threshold to avoid rejecting real license faces.
static CGFloat const kMinFaceSizeRatio = 0.002;

// Maximum face area relative to image area
static CGFloat const kMaxFaceSizeRatio = 0.70;

@interface FaceDetectorHelper ()

@property (nonatomic, strong) dispatch_queue_t detectionQueue;
@property (atomic, assign) BOOL isBusy;

@end

@implementation FaceDetectorHelper

#pragma mark - Initialization

- (instancetype)init {
    self = [super init];
    if (self) {
        _detectionQueue = dispatch_queue_create("com.sos.facedetector", DISPATCH_QUEUE_SERIAL);
    }
    return self;
}

#pragma mark - Public Methods

- (void)detectFaceInImage:(UIImage *)image {
    if (!image) {
        NSLog(@"[FaceDetector] Invalid image provided");
        [self notifyNoFaceDetected];
        return;
    }

    dispatch_async(self.detectionQueue, ^{
        [self performFaceDetection:image];
    });
}

- (void)checkFacePresenceInPixelBuffer:(CVPixelBufferRef)pixelBuffer {
    if (self.isBusy || !pixelBuffer) return;
    self.isBusy = YES;

    // Retain buffer for async use on the detection queue
    CVPixelBufferRetain(pixelBuffer);
    dispatch_async(self.detectionQueue, ^{
        [self performPresenceDetectionInPixelBuffer:pixelBuffer];
        CVPixelBufferRelease(pixelBuffer);
    });
}

/**
 * Runs VNDetectFaceRectanglesRequest on a pixel buffer.
 * Reports presence via delegate if a face passes size filters.
 * Orientation: kCGImagePropertyOrientationRight = phone held in portrait (camera buffer is landscape-right).
 */
- (void)performPresenceDetectionInPixelBuffer:(CVPixelBufferRef)pixelBuffer {
    __weak __typeof(self) weakSelf = self;

    VNDetectFaceRectanglesRequest *request = [[VNDetectFaceRectanglesRequest alloc]
        initWithCompletionHandler:^(VNRequest *req, NSError *error) {
            __strong __typeof(weakSelf) strongSelf = weakSelf;
            if (!strongSelf) return;
            strongSelf.isBusy = NO;

            if (error || !req.results || req.results.count == 0) return;

            // Check that at least one face passes the area-size filter
            for (VNFaceObservation *face in req.results) {
                CGFloat area = face.boundingBox.size.width * face.boundingBox.size.height;
                if (area >= kMinFaceSizeRatio && area <= kMaxFaceSizeRatio) {
                    id<FaceDetectorHelperDelegate> delegate = strongSelf.delegate;
                    if (delegate && [delegate respondsToSelector:@selector(faceDetectorHelperDidDetectFacePresence:)]) {
                        dispatch_async(dispatch_get_main_queue(), ^{
                            [delegate faceDetectorHelperDidDetectFacePresence:strongSelf];
                        });
                    }
                    return;
                }
            }
        }];

    // Phone held in portrait → video buffer is landscape-right (top of physical scene is on the right of the buffer)
    VNImageRequestHandler *handler = [[VNImageRequestHandler alloc]
        initWithCVPixelBuffer:pixelBuffer
                  orientation:kCGImagePropertyOrientationRight
                      options:@{}];

    NSError *handlerError = nil;
    if (![handler performRequests:@[request] error:&handlerError]) {
        self.isBusy = NO;
        NSLog(@"[FaceDetector] Pixel buffer detection error: %@", handlerError);
    }
}

#pragma mark - Orientation Normalization

- (UIImage *)normalizeImageOrientation:(UIImage *)image {
    if (image.imageOrientation == UIImageOrientationUp) {
        return image;
    }

    // Draw into a new context to apply the orientation transform.
    // This creates a new CGImage with pixels in the correct display orientation,
    // so CGImage dimensions match UIImage.size and Vision bounding boxes align
    // with the pixel data used by CGImageCreateWithImageInRect.
    UIGraphicsBeginImageContextWithOptions(image.size, NO, image.scale);
    [image drawInRect:CGRectMake(0, 0, image.size.width, image.size.height)];
    UIImage *normalized = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();

    NSLog(@"[FaceDetector] Normalized image orientation from %ld to Up",
          (long)image.imageOrientation);

    return normalized ?: image;
}

#pragma mark - Face Detection

- (void)performFaceDetection:(UIImage *)image {
    // Normalize orientation so CGImage pixels match UIImage display coordinates.
    // Without this, Vision bounding boxes (relative to raw CGImage) don't match
    // the crop coordinates (computed from UIImage.size), causing wrong region extraction.
    UIImage *normalizedImage = [self normalizeImageOrientation:image];

    CGImageRef cgImage = normalizedImage.CGImage;
    if (!cgImage) {
        NSLog(@"[FaceDetector] Could not get CGImage from UIImage");
        [self notifyNoFaceDetected];
        return;
    }

    size_t cgWidth = CGImageGetWidth(cgImage);
    size_t cgHeight = CGImageGetHeight(cgImage);

    NSLog(@"[FaceDetector] Analyzing image: UIImage size=%@, CGImage=%zux%zu, orientation=%ld",
          NSStringFromCGSize(normalizedImage.size), cgWidth, cgHeight,
          (long)normalizedImage.imageOrientation);

    // Use VNDetectFaceLandmarksRequest (superset of rectangles) so that the
    // VNFaceObservation.roll property is reliably populated — required for tilt correction.
    VNDetectFaceLandmarksRequest *request = [[VNDetectFaceLandmarksRequest alloc]
        initWithCompletionHandler:^(VNRequest *request, NSError *error) {
            if (error) {
                NSLog(@"[FaceDetector] Detection error: %@", error);
                [self notifyError:error];
                return;
            }

            [self handleDetectionResults:request.results forImage:normalizedImage];
        }];

    // No orientation parameter needed since the image is already normalized (orientation=Up).
    // Vision and CGImageCreateWithImageInRect now operate in the same coordinate space.
    VNImageRequestHandler *handler = [[VNImageRequestHandler alloc]
        initWithCGImage:cgImage
                options:@{}];

    NSError *error = nil;
    if (![handler performRequests:@[request] error:&error]) {
        NSLog(@"[FaceDetector] Failed to perform request: %@", error);
        [self notifyError:error];
    }
}

- (void)handleDetectionResults:(NSArray<VNFaceObservation *> *)results forImage:(UIImage *)image {
    if (!results || results.count == 0) {
        NSLog(@"[FaceDetector] No faces detected in image");
        [self notifyNoFaceDetected];
        return;
    }

    NSLog(@"[FaceDetector] Detected %lu face(s)", (unsigned long)results.count);

    // Log all detected faces for debugging
    CGFloat imageArea = image.size.width * image.size.height;
    for (NSUInteger i = 0; i < results.count; i++) {
        VNFaceObservation *face = results[i];
        CGRect bbox = face.boundingBox;
        CGFloat faceArea = (bbox.size.width * image.size.width) * (bbox.size.height * image.size.height);
        CGFloat areaRatio = faceArea / imageArea;
        NSLog(@"[FaceDetector] Face %lu: bbox=(%.3f, %.3f, %.3f, %.3f) confidence=%.2f areaRatio=%.4f",
              (unsigned long)i, bbox.origin.x, bbox.origin.y,
              bbox.size.width, bbox.size.height,
              face.confidence, areaRatio);
    }

    // Find the best face
    VNFaceObservation *bestFace = [self findBestFaceInResults:results forImage:image];

    if (!bestFace) {
        NSLog(@"[FaceDetector] No suitable face found (all rejected by size filter)");
        [self notifyNoFaceDetected];
        return;
    }

    // Extract face region from the normalized image
    UIImage *croppedFace = [self extractFaceRegion:bestFace fromImage:image];

    if (croppedFace) {
        NSLog(@"[FaceDetector] Successfully extracted face portrait, size=%@",
              NSStringFromCGSize(croppedFace.size));
        [self notifyFaceDetected:croppedFace];
    } else {
        NSLog(@"[FaceDetector] Failed to crop face region");
        [self notifyNoFaceDetected];
    }
}

- (VNFaceObservation *)findBestFaceInResults:(NSArray<VNFaceObservation *> *)results
                                    forImage:(UIImage *)image {
    CGFloat imageWidth = image.size.width;
    CGFloat imageHeight = image.size.height;
    CGFloat imageArea = imageWidth * imageHeight;

    VNFaceObservation *bestFace = nil;
    CGFloat bestScore = -1;

    for (VNFaceObservation *face in results) {
        CGRect normalizedBounds = face.boundingBox;

        // Calculate face area ratio
        CGFloat faceWidth = normalizedBounds.size.width * imageWidth;
        CGFloat faceHeight = normalizedBounds.size.height * imageHeight;
        CGFloat faceArea = faceWidth * faceHeight;
        CGFloat areaRatio = faceArea / imageArea;

        // Skip faces outside valid size range
        if (areaRatio < kMinFaceSizeRatio) {
            NSLog(@"[FaceDetector] Face rejected: too small (ratio=%.4f, min=%.4f)",
                  areaRatio, kMinFaceSizeRatio);
            continue;
        }
        if (areaRatio > kMaxFaceSizeRatio) {
            NSLog(@"[FaceDetector] Face rejected: too large (ratio=%.4f, max=%.4f)",
                  areaRatio, kMaxFaceSizeRatio);
            continue;
        }

        // Score: prioritize larger faces (the license face should be the most prominent).
        // Size accounts for 70% of the score. Position accounts for 30%.
        CGFloat sizeScore = areaRatio;

        // Position scoring: US driver licenses typically have the photo on the left side.
        // In a portrait camera frame, the left side of the license (held horizontally)
        // maps to the left-center of the frame.
        CGFloat faceCenterX = normalizedBounds.origin.x + normalizedBounds.size.width / 2.0;
        CGFloat faceCenterY = normalizedBounds.origin.y + normalizedBounds.size.height / 2.0;

        // Prefer faces in the left-center area of the frame
        // Vision coordinate system: (0,0) = bottom-left, (1,1) = top-right
        CGFloat positionScore = 1.0 - (fabs(faceCenterX - 0.35) + fabs(faceCenterY - 0.5)) / 2.0;
        positionScore = MAX(0, positionScore);

        CGFloat totalScore = sizeScore * 0.7 + positionScore * 0.3;

        NSLog(@"[FaceDetector] Face score: size=%.4f pos=%.4f total=%.4f (center=%.2f,%.2f)",
              sizeScore, positionScore, totalScore, faceCenterX, faceCenterY);

        if (totalScore > bestScore) {
            bestScore = totalScore;
            bestFace = face;
        }
    }

    if (bestFace) {
        NSLog(@"[FaceDetector] Selected face with score=%.4f", bestScore);
    }

    return bestFace;
}

- (UIImage *)extractFaceRegion:(VNFaceObservation *)face fromImage:(UIImage *)image {
    CGFloat imageWidth  = image.size.width;
    CGFloat imageHeight = image.size.height;

    // Convert Vision coords (bottom-left, normalized) → UIKit pixels (top-left)
    CGRect nb = face.boundingBox;
    CGFloat faceX = nb.origin.x * imageWidth;
    CGFloat faceY = (1.0 - nb.origin.y - nb.size.height) * imageHeight;
    CGFloat faceW = nb.size.width  * imageWidth;
    CGFloat faceH = nb.size.height * imageHeight;

    // Face centre in image pixels
    CGFloat cx = faceX + faceW / 2.0;
    CGFloat cy = faceY + faceH / 2.0;

    NSLog(@"[FaceDetector] Face: x=%.0f y=%.0f w=%.0f h=%.0f centre=(%.0f,%.0f)",
          faceX, faceY, faceW, faceH, cx, cy);

    // Roll angle (0 if not available)
    CGFloat roll = (face.roll != nil) ? face.roll.floatValue : 0.0;
    NSLog(@"[FaceDetector] Roll = %.3f rad (%.1f°)", roll, roll * 180.0 / M_PI);

    // Desired output square side: face size + portrait padding on all sides
    CGFloat faceSize   = MAX(faceW, faceH);
    CGFloat outputSize = ceil(faceSize * (1.0 + kFacePaddingFactor * 2.0));

    // Pre-rotation crop must be large enough so rotating it by `roll` leaves
    // no empty corners inside the central outputSize × outputSize region.
    // Required side = outputSize × (|cos| + |sin|), i.e. the bounding box of
    // a rotated square of side outputSize.
    CGFloat sinA = fabs(sin(roll));
    CGFloat cosA = fabs(cos(roll));
    CGFloat preCropSize = ceil(outputSize * (cosA + sinA)) + 4;

    // Crop centred on the face, clamped to image bounds
    CGFloat left   = MAX(0,           cx - preCropSize / 2.0);
    CGFloat top    = MAX(0,           cy - preCropSize / 2.0);
    CGFloat right  = MIN(imageWidth,  cx + preCropSize / 2.0);
    CGFloat bottom = MIN(imageHeight, cy + preCropSize / 2.0);

    CGRect preCropRect = CGRectMake(left, top, right - left, bottom - top);
    if (preCropRect.size.width <= 0 || preCropRect.size.height <= 0) {
        NSLog(@"[FaceDetector] Invalid pre-crop rect");
        return nil;
    }

    CGImageRef cgLarge = CGImageCreateWithImageInRect(image.CGImage, preCropRect);
    if (!cgLarge) {
        NSLog(@"[FaceDetector] CGImageCreateWithImageInRect (large) failed");
        return nil;
    }
    UIImage *largeImage = [UIImage imageWithCGImage:cgLarge
                                              scale:image.scale
                                        orientation:UIImageOrientationUp];
    CGImageRelease(cgLarge);

    // Rotate the large image to align the face upright.
    // UIKit image context has Y-axis flipped: positive angle = clockwise in output.
    // roll > 0 = face tilted counter-clockwise → clockwise correction = +roll.
    UIImage *rotated = (fabs(roll) > 0.052)
        ? [self rotateImage:largeImage byRadians:roll]
        : largeImage;

    // Crop a square from the centre of the rotated image — the face is centred
    // there and all corners are filled with real image content (no white gaps).
    return [self centerSquareCrop:rotated size:outputSize];
}

/**
 * Rotates an image by `radians` around its centre.
 * Canvas expands to contain the full rotated content (no clipping).
 */
- (UIImage *)rotateImage:(UIImage *)image byRadians:(CGFloat)radians {
    CGFloat sinA = fabs(sin(radians));
    CGFloat cosA = fabs(cos(radians));
    CGFloat newW = image.size.width  * cosA + image.size.height * sinA;
    CGFloat newH = image.size.width  * sinA + image.size.height * cosA;

    UIGraphicsBeginImageContextWithOptions(CGSizeMake(newW, newH), NO, image.scale);
    CGContextRef ctx = UIGraphicsGetCurrentContext();
    CGContextTranslateCTM(ctx, newW / 2.0, newH / 2.0);
    CGContextRotateCTM(ctx, radians);
    [image drawInRect:CGRectMake(-image.size.width  / 2.0,
                                 -image.size.height / 2.0,
                                  image.size.width,
                                  image.size.height)];
    UIImage *result = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    return result ?: image;
}

/**
 * Crops a square of `size` pixels from the centre of `image`.
 * The centre of the image = the (corrected) face centre.
 */
- (UIImage *)centerSquareCrop:(UIImage *)image size:(CGFloat)size {
    CGFloat side = MIN(size, MIN(image.size.width, image.size.height));
    CGRect cropRect = CGRectMake((image.size.width  - side) / 2.0,
                                 (image.size.height - side) / 2.0,
                                  side, side);
    CGImageRef cgImage = CGImageCreateWithImageInRect(image.CGImage, cropRect);
    if (!cgImage) return image;
    UIImage *result = [UIImage imageWithCGImage:cgImage
                                          scale:image.scale
                                    orientation:UIImageOrientationUp];
    CGImageRelease(cgImage);
    return result;
}

#pragma mark - Delegate Notifications

- (void)notifyFaceDetected:(UIImage *)croppedFace {
    dispatch_async(dispatch_get_main_queue(), ^{
        [self.delegate faceDetectorHelper:self didDetectFace:croppedFace];
    });
}

- (void)notifyNoFaceDetected {
    dispatch_async(dispatch_get_main_queue(), ^{
        [self.delegate faceDetectorHelperDidNotDetectFace:self];
    });
}

- (void)notifyError:(NSError *)error {
    dispatch_async(dispatch_get_main_queue(), ^{
        if ([self.delegate respondsToSelector:@selector(faceDetectorHelper:didFailWithError:)]) {
            [self.delegate faceDetectorHelper:self didFailWithError:error];
        } else {
            [self.delegate faceDetectorHelperDidNotDetectFace:self];
        }
    });
}

@end
